File: frontend/src/core/api/clipboard.ts
/**
 * Roaming CLIPBOARD HISTORY API (`/api/v1/clipboard`).
 *
 * The browser clipboard holds one item, is not shared between tabs and is lost
 * on reload. This history keeps the recent clips server-side, per user, as the
 * same `KubunoDataEnvelope` the `core.data-card` renderers already understand —
 * so a copy made in the spreadsheet can be pasted from the chat, from another
 * tab, or from another device.
 *
 * The backend owns the rules (dedup by content, 30 unpinned entries kept,
 * 256 KB payload cap): the client only pushes and reads.
 */
import { api } from './client'
import type { KubunoDataEnvelope } from '../registry/DataTransferRegistry'

export interface ClipboardItem {
  id: string
  /** Producing module id ('office', 'maps'…). */
  module: string
  /** Envelope type ('office.shape', 'maps.place'…). */
  kind: string
  title: string | null
  /** Human-readable summary shown in the pane. */
  preview: string | null
  /** The envelope itself — rendered through the producer's data-card renderer. */
  payload: KubunoDataEnvelope
  href: string | null
  /** Pinned entries survive the trim and « Effacer l'historique ». */
  pinned: boolean
  created_at: string
  updated_at: string
}

export const clipboardApi = {
  async list(limit = 30): Promise<ClipboardItem[]> {
    const { data } = await api.get<{ items: ClipboardItem[] }>('/clipboard', { params: { limit } })
    return data.items ?? []
  },

  /**
   * Record a clip. Never throws at the caller: a history that is unavailable
   * (offline, older server) must not break the copy itself, which has already
   * happened on the system clipboard.
   */
  async push(envelope: KubunoDataEnvelope, opts?: { pinned?: boolean }): Promise<ClipboardItem | null> {
    try {
      const { data } = await api.post<{ item: ClipboardItem }>('/clipboard', {
        module: envelope.module,
        kind: envelope.type,
        title: envelope.title ?? null,
        preview: envelope.text ?? null,
        payload: envelope,
        href: envelope.href ?? null,
        pinned: opts?.pinned ?? false,
      })
      return data.item ?? null
    } catch {
      return null
    }
  },

  async setPinned(id: string, pinned: boolean): Promise<void> {
    await api.patch(`/clipboard/${id}`, { pinned })
  },

  async remove(id: string): Promise<void> {
    await api.delete(`/clipboard/${id}`)
  },

  /** Clears everything except the pinned entries. */
  async clear(): Promise<number> {
    const { data } = await api.delete<{ deleted: number }>('/clipboard')
    return data.deleted ?? 0
  },
}

File: crates/kubuno-core/src/models/clipboard.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use uuid::Uuid;
use validator::Validate;

/// One entry of a user's roaming clipboard history.
///
/// `payload` is the cross-module JSON envelope the producing module copied, so
/// the frontend renders history entries through the very same `core.data-card`
/// renderers it uses when pasting.
#[derive(Debug, Clone, Serialize, FromRow)]
pub struct ClipboardItem {
    pub id:         Uuid,
    pub module:     String,
    pub kind:       String,
    pub title:      Option<String>,
    pub preview:    Option<String>,
    pub payload:    serde_json::Value,
    pub href:       Option<String>,
    pub pinned:     bool,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// Push a clip. The server computes the fingerprint (dedup) and enforces the
/// size cap — a client is never trusted with either.
#[derive(Debug, Deserialize, Validate)]
pub struct PushClipboardDto {
    #[validate(length(min = 1, max = 100, message = "Module: 1-100 caractères"))]
    pub module:  String,
    #[validate(length(min = 1, max = 100, message = "Type: 1-100 caractères"))]
    pub kind:    String,
    #[validate(length(max = 500))]
    pub title:   Option<String>,
    /// Human-readable summary; truncated server-side rather than rejected, so a
    /// long copy never fails for a cosmetic reason.
    pub preview: Option<String>,
    pub payload: serde_json::Value,
    #[validate(length(max = 1000))]
    pub href:    Option<String>,
    /// Pin the entry right away (rare; the pane's pin button is the usual path).
    pub pinned:  Option<bool>,
}

#[derive(Debug, Deserialize, Validate)]
pub struct UpdateClipboardDto {
    pub pinned: Option<bool>,
}
